File: tools/dice_combination.js
function calcularValoresDados(combinacao) {
    let minimo = combinacao.reduce((acc, dado) => acc + parseInt(dado.substring(1)), 0);
    let maximo = combinacao.reduce((acc, dado) => acc + parseInt(dado.substring(1)) * (Math.floor(parseInt(dado.substring(1)) / 2) + 1), 0);
    return { minimo, maximo };
}

function gerarCombinacoes(dadosDisponiveis, tamanho) {
    if (tamanho === 1) {
        return dadosDisponiveis.map(dado => [dado]);
    }
    const combinacoes = [];
    dadosDisponiveis.forEach((dadoAtual, index) => {
        const combinacoesMenores = gerarCombinacoes(dadosDisponiveis.slice(index), tamanho - 1);
        combinacoesMenores.forEach(combinacao => {
            combinacoes.push([dadoAtual].concat(combinacao));
        });
    });
    return combinacoes;
}

function possiveisFormulas(metaDano, dadosDisponiveis) {
    let formulasValidas = new Set();

    // Adicionando combinações de um único dado e um valor fixo
    dadosDisponiveis.forEach(dado => {
        let valorDado = parseInt(dado.substring(1));
        if (valorDado <= metaDano) {
            let valorFixo = metaDano - valorDado;
            let formula = dado + (valorFixo > 0 ? " + " + valorFixo : "");
            formulasValidas.add(formula);
        }
    });

    // Adicionando combinações de dois dados e um valor fixo
    dadosDisponiveis.forEach(dado1 => {
        dadosDisponiveis.forEach(dado2 => {
            let valorDado1 = parseInt(dado1.substring(1));
            let valorDado2 = parseInt(dado2.substring(1));
            let minimo = valorDado1 + valorDado2;
            if (minimo <= metaDano) {
                let valorFixo = metaDano - minimo;
                let formula = dado1 + " + " + dado2 + (valorFixo > 0 ? " + " + valorFixo : "");
                formulasValidas.add(formula);
            }
        });
    });

    // Adicionando combinações de múltiplos dados
    for (let i = 3; i <= 5; i++) {
        let combinacoes = gerarCombinacoes(dadosDisponiveis, i);
        combinacoes.forEach(combinacao => {
            const { minimo, maximo } = calcularValoresDados(combinacao);
            if (minimo <= metaDano && metaDano <= maximo) {
                const valorFixo = metaDano - minimo;
                const formula = combinacao.join(" + ") + (valorFixo > 0 ? " + " + valorFixo : "");
                formulasValidas.add(formula);
            }
        });
    }

    return Array.from(formulasValidas).sort();
}

// Exemplo de uso
const dadosDisponiveis = ['d4', 'd6', 'd8', 'd10', 'd12'];
const metaDano = 60;
const formulas = possiveisFormulas(metaDano, dadosDisponiveis);
console.log(formulas);

File: tools/dice_combination.rb
require 'set'

# Função para calcular o valor mínimo e máximo de uma combinação de dados
def calcular_valores_dados(combinacao)
  minimo = combinacao.sum { |dado| dado[1..].to_i }
  maximo = combinacao.sum { |dado| dado[1..].to_i * (dado[1..].to_i / 2 + 1) }
  return minimo, maximo
end

def possiveis_formulas(meta_dano, dados_disponiveis)
  # Calculando todas as combinações possíveis de dados, considerando a ordem irrelevante
  combinacoes_dados = []
  (1..5).each do |i|
    combinacoes_dados += dados_disponiveis.repeated_combination(i).to_a
  end

  # Encontrando combinações que podem atingir a meta de dano
  formulas_validas = Set.new
  
  combinacoes_dados.each do |combinacao|
    minimo, maximo = calcular_valores_dados(combinacao)
    if minimo <= meta_dano && meta_dano <= maximo
      valor_fixo = meta_dano - minimo
      formula = combinacao.join(' + ') + (valor_fixo > 0 ? " + #{valor_fixo}" : '')
      formulas_validas.add(formula)
    end
  end

  # Adicionando combinações com um único dado e um valor fixo
  dados_disponiveis.each do |dado|
    dado_minimo = dado[1..].to_i
    if dado_minimo <= meta_dano
      valor_fixo = meta_dano - dado_minimo
      formula = dado + (valor_fixo > 0 ? " + #{valor_fixo}" : '')
      formulas_validas.add(formula)
    end
  end

  # Ordenando e removendo duplicatas
  formulas_validas.to_a.sort
end

# Mostrando todas as fórmulas válidas
possiveis_formulas(ARGV[0].to_i, ['d4', 'd6', 'd8', 'd10', 'd12']).each { |formula| puts formula }
